### ISSUE TITLE
GRPO Improvement checklist

### ISSUE THREAD

**Issue by RedTachyon** [pytorch/torchtune#2421]
It's alive! It's alive! (#2326)

The recipe definitely works (as in, I can run it and reach like a 60% success rate on GSM8k with a 3B model), but it's somewhat barebones and underoptimized. Here, I want to keep track of all the most important features and improvements that I think are missing. I'll probably go through this at some point, at some pace, but if anyone else wants to contribute - you can grab something from this list.

## Improvements
- [ ] Figure out how to move things from dev into the main repository (decide on the final APIs etc.)
- [ ] A proper eval workflow, so that every once in a while we run a full eval on the test set. Alternatively, a separate evaluation recipe? (note: I actually have a working separate eval recipe that works with the same paradigm as GRPO, goes through the full test dataset, computes success/reward and saves it to a file inside the checkpoint - happy to make it a PR)
- [ ] Adding proper (unit) tests, compliant with the normal torchtune testing workflow
- [ ] Adding proper documentation to everything
- [ ] Refactoring of the GRPO losses (there's also a research-y question of what loss should be used - there's some ambiguity in the paper and reference implementations)
- [ ] More modular approach to reward computation (probably as a component with the regular OmegaConf setup)
- [ ] Step-based checkpointing (#2384) - this is pretty important, since one epoch can be very long, leading to very infrequent checkpoints
- [ ] Memory profiling and experiments on "controlled" hardware (a recipe tuned to work ~optimally on a node of 8xH100, or on a single H100, or on smaller hardware with e.g. LoRA)
- [ ] Optimization of the default recipe - maybe we can get a big performance boost e.g. by doing `ppo_epochs>1`?
- [ ] Dataset improvements (gsm8k is functional but could use some polish, there's also the MATH and DeepscaleR datasets with a similar structure that can be added)
- [ ] A single-device version - should be pretty simple, but probably also slow. Might require gradient accumulation to properly work, I tend to get bad results with small batch sizes.
- [ ] Try to improve generation speed by using vLLM (or something else)?
- [ ] Probably more to be found soon

## Bugs
Because of course I found a bug right after everything was finalized. There will likely be more, so this subsection might or might not be useful.
- [ ] Right now, [generate_trajectory_batched](https://github.com/pytorch/torchtune/blob/cf0142bd2eb5eabe33a3d43eca4f07ec48a8749f/recipes/dev/grpo_full_finetune_distributed.py#L816) can crash when the different generations are of different size. For example, one batch of completions generated the full 512 tokens, but another one got truncated at 300 because it hit a stop token everywhere. So you have tensors of shapes [16, 512] and [16, 300], and try to concatenate them across zero-th axis - which obviously doesn't work. The tensors need to be padded to consistent length.

---

Note to maintainers - I took the liberty to create this centralized checklist since I still have all the necessary improvements in my context window. In principle, each bullet point could be a separate issue, but that would probably be a nightmare. We can coordinate the effort around this issue, and start adding the improvements, one PR at a time.

**Comment by krammnic** [comment]
UPD: MATH dataset from "EleutherAI/hendrycks_math" is added, but it still might be banned in future.

**Comment by ebsmothers** [comment]
Thank you so much for creating this checklist @RedTachyon! It's great to have all these items in one place. Actually a couple of the improvements you listed are horizontal changes we've been wanting to enable across the repo anyways -- I'm thinking specifically of vLLM, eval datasets, and step-based checkpointing. Step-based checkpointing is already in progress and I think some basic eval shouldn't be too hard (I think #2238 was going in that direction and may just need some minor changes). Proper vLLM integration may be a bigger effort, but we are planning to get going on this asap. And thanks @krammnic for already working on a couple of these!

**Comment by emattia** [comment]
Hey @RedTachyon, excellent work! I am actively testing your scripts, have reproduced using the default gsm8k dataset, and would like to contribute along the lines of this issue, and sanity check my results. Our team has access to a large set of H100 VMs right now, and next week I'll be diving into this full time.

Here is a list of my motivations, it would be great to discuss how I can help you achieve your goals:
- I could work on "controlled" hardware implementation for 8x H100s, and possibly AWS EC2s our customers use. 
- I'd like to run some HPO experiments, for example over ppo_epochs and other RL hyperparams. It'd be helpful to understand more details how you are thinking about parameter tuning so I can launch a well-designed grid search using our spare capacity.  
- A big winner for us would be if we could turn around a multinode example that post-trains a large (≥70B) base model. 
- Step-based checkpointing would also be a helpful pattern for us.
